### ISSUE TITLE
deploy/rbd/kubernetes: service definitions unnecessary?

### ISSUE THREAD

**Issue by pohly** [ceph/ceph-csi#81]
Why do `deploy/rbd/kubernetes/csi-rbdplugin-attacher.yaml` and `deploy/rbd/kubernetes/csi-rbdplugin-provisioner.yaml` contain `Service` definitions?

Even if there is a reason for them, I'd prefer to have them in separate files, because then admins can decide whether they want them and/or create/delete them independently.

**Comment by gman0** [comment]
Please see the paragraph about the headless service requirement at https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#limitations

**Comment by pohly** [comment]
> StatefulSets currently require a Headless Service to be responsible for the network identity of the Pods.  You are responsible for creating this Service.

You probably mean that, right? The way I read this is that "if you need a network identity, then you must create a Service". But for provisioner and attacher we don't need any network identity, do we?

Let me rephrase: what breaks if the Services get removed from the .yaml files?

**Comment by rootfs** [comment]
@pohly this is a requirement/limitation

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#statefulset-v1-apps
```
serviceNamestring | serviceName is the name of the service that governs this StatefulSet. This service must exist before the StatefulSet, and is responsible for the network identity of the set. Pods get DNS/hostnames that follow the pattern: pod-specific-string.serviceName.default.svc.cluster.local where "pod-specific-string" is managed by the StatefulSet controller.

```

**Comment by pohly** [comment]
I asked on the wg-csi Slack channel and Michelle confirmed that the Service isn't needed. She provided these two examples:
https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/deploy/kubernetes/stable/controller.yaml
https://github.com/kubernetes-sigs/gcp-filestore-csi-driver/blob/master/deploy/kubernetes/manifests/controller.yaml

Can we reopen this issue? I can test the removal and provide a PR.

**Comment by rootfs** [comment]
Thanks getting the feedback, I cannot speak for GCP's case but before we have a definitive answer from upstream k8s sig, I prefer keeping them here. The headless service doesn't consume networking resources. But this is surely confusing. Would you please opening a doc PR and get this limitation into details?

**Comment by pohly** [comment]
With doc PR, do you mean against https://github.com/kubernetes/kubernetes/tree/master/docs? I'm not sure at the moment where the source code for the two quotes above is.

**Comment by rootfs** [comment]
let's do it here 
https://github.com/ceph/ceph-csi/blob/master/docs/deploy-rbd.md

**Comment by pohly** [comment]
@rootfs I don't know whether this is what you wanted, but I've opened PR #83 to remove the Service definitions. I can confirm that provisioning and attaching works fine this way.
